cpp/src/arrow/array/builder_binary.h

```diff
@@ -51,10 +51,12 @@ class BaseBinaryBuilder : public ArrayBuilder {
   using offset_type = typename TypeClass::offset_type;
 
   explicit BaseBinaryBuilder(MemoryPool* pool = default_memory_pool())
-      : ArrayBuilder(pool), offsets_builder_(pool), value_data_builder_(pool) {}
+      : ArrayBuilder(pool), offsets_builder_(pool),
+        value_data_builder_(pool), type_(binary()) {}
```

Comment:
Please delete this constructor; I think it's more understandable if all subclasses of BaseBinaryBuilder are required to pass the resulting type

Reply:
Done